src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.Worker.cs

```diff
@@ -0,0 +1,496 @@
+            {
+                var results = ImmutableArray.CreateBuilder<DocumentInfo>();
+
+                foreach (var info in documentFileInfos)
+                {
+                    GetDocumentNameAndFolders(info.LogicalPath, out var name, out var folders);
+
+                    var documentInfo = DocumentInfo.Create(
+                        DocumentId.CreateNewId(projectId, debugName: info.FilePath),
+                        name,
+                        folders,
+                        info.SourceCodeKind,
+                        new FileTextLoader(info.FilePath, encoding),
+                        info.FilePath,
+                        info.IsGenerated);
+
+                    results.Add(documentInfo);
+                }
+
+                return results.ToImmutable();
+            }
+
+            private static readonly char[] s_directorySplitChars = new char[] { Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar };
+
+            private static void GetDocumentNameAndFolders(string logicalPath, out string name, out ImmutableArray<string> folders)
+            {
+                var pathNames = logicalPath.Split(s_directorySplitChars, StringSplitOptions.RemoveEmptyEntries);
+                if (pathNames.Length > 0)
+                {
+                    if (pathNames.Length > 1)
+                    {
+                        folders = pathNames.Take(pathNames.Length - 1).ToImmutableArray();
+                    }
+                    else
+                    {
+                        folders = ImmutableArray<string>.Empty;
+                    }
+
+                    name = pathNames[pathNames.Length - 1];
+                }
+                else
+                {
+                    name = logicalPath;
+                    folders = ImmutableArray<string>.Empty;
+                }
+            }
+
+            private void CheckForDuplicateDocuments(ImmutableArray<DocumentInfo> documents, ImmutableArray<DocumentInfo> additionalDocuments, string projectFilePath, ProjectId projectId)
+            {
+                var paths = new HashSet<string>(PathUtilities.Comparer);
+                foreach (var doc in documents.Concat(additionalDocuments))
+                {
+                    if (paths.Contains(doc.FilePath))
+                    {
+                        var message = string.Format(WorkspacesResources.Duplicate_source_file_0_in_project_1, doc.FilePath, projectFilePath);
+                        var diagnostic = new ProjectDiagnostic(WorkspaceDiagnosticKind.Warning, message, projectId);
+
+                        _diagnosticReporter.Report(diagnostic);
+                    }
+
+                    paths.Add(doc.FilePath);
+                }
+            }
+
+            private bool IsSupportedLanguage(string languageName)
+                => _workspace.Services.SupportedLanguages.Contains(languageName);
```

Comment:
I don't think this is useful any longer. The only place it's called is in an assert that fires _long_ after it would have been relevant. I'll just remove it.